system/lib/wasmfs/backends/memory_backend.cpp

```diff
@@ -38,40 +38,12 @@ MemoryDirectory::findEntry(const std::string& name) {
   });
 }
 
-std::shared_ptr<File> MemoryDirectory::getChild(const std::string& name) {
-  if (auto entry = findEntry(name); entry != entries.end()) {
-    return entry->child;
-  }
-  return nullptr;
-}
-
 bool MemoryDirectory::removeChild(const std::string& name) {
   auto entry = findEntry(name);
-  if (entry == entries.end()) {
-    return false;
-  }
-  entries.erase(entry);
-  return true;
-}
-
-std::shared_ptr<File> MemoryDirectory::insertChild(const std::string& name,
-                                                   std::shared_ptr<File> file) {
-  if (auto entry = findEntry(name); entry != entries.end()) {
-    return entry->child;
-  }
-  entries.push_back({name, file});
-  return file;
-}
-
-std::string MemoryDirectory::getName(std::shared_ptr<File> file) {
-  auto entry =
-    std::find_if(entries.begin(), entries.end(), [&](const auto& entry) {
-      return entry.child == file;
-    });
   if (entry != entries.end()) {
-    return entry->name;
+    entries.erase(entry);
   }
-  return {};
+  return true;
```

Comment:
Yeah, originally I wanted `MemoryFile` to do nothing like you're thinking. Unfortunately the `getEntries()` method makes that difficult. Since it can't assume all the entries are already in the cache, it has to unconditionally query the backend for the directory entries, which means that the backend needs to know what the entries are. Unfortunately that means `MemoryDirectory` cannot completely abdicate responsibility for keeping track of the entries.

Reply:
> Since it can't assume all the entries are already in the cache

Hmm, why can't it assume that for `MemoryBackend`?

My understanding was that the dcache is all there is in that backend - there are no extra things not tracked in the dcache. So `MemoryBackend::getEntries()` can return an empty set, etc.

Reply:
As a special case, it could assume that all entries are in the cache for MemoryBackend, but there is no way to cleanly implement that special case. In the general case across all backends it cannot assume all entries are in the cache because a persistent backend may have entries that have never been requested before.

Reply:
True, but can't we remove `std::vector<ChildEntry> entries;` from `MemoryDirectory` and then not implement anything? If the default behavior of insert/remove/etc. is to use the dcache, and if not in the dcache to look in the child, then the memory backend would simply do nothing in the latter hooks. And then the dcache would be all there is there.

Reply:
How would you implement `getEntries()` if you did that?

Here's the logic for arriving at the current implementation of `getEntries`, considering a general persistent backend:
1. We can't know that the cache already contains all the entries, so we have to call out to the backend in `getEntries()`.
2. The backend doesn't know what's in the cache, so it returns all the entries.
3. The backend can't have included the mount points, so we append them to the entries returned by the backend. We don't append any other cached entries because the backend would already have included them.

The standard contract of `Directory::getEntries()` is that the backend returns _all_ the entries. The in-memory backend has to do that as well; it doesn't get to be special.

Reply:
I agree `getEntries()` should return all the entries. I would imagine it could work like this: it makes a list of the dcache entries. It then appends to that list the results of a call to the backend's `getExtraEntries()` or something like that - where that function only adds "extra" things not in the dcache. In other words, `Directory::getEntries()` would not be virtual, and would call a backend hook that is. Does that make sense? I'm not opposed to calling out to the backend, I just expect to see the backend only add the backend-specific stuff, and so in particular MemoryBackend would have nothing to add.

Reply:
Having something like `getExtraEntries()` would require every other backend to keep track of which entries are already in the dcache, which is not a good trade off. Another option would be to have `Directory::Handle::getEntries` create entries from the dcache then query all the results from the backend and deduplicate the results. I'm not a huge fan of that either because it's a bunch of extra work that is only useful for the in-memory backend.

Reply:
Ah, I see what you mean, when something moved into the dcache it would need to be removed from the "non-dcache" list. That does sound like not the best thing.